G-API: oneVPL merge DX11 acceleration

sivanov-work opened the pull request:
Add DX11 allocation resources: surfaces
Implement actual DX11 acceleration with MFX allocator
Reuse IDeviceSelector interface for selecting device in DX11 accelerator policy

Add `shared_lock` & `elastics_barrier` to work out R-W pattern & provide lock-free access to MediaFrame data.
Add unit tests for lock-free primitives

Improve decode pipeline: add yield on queue decode operations for used up busy surfaces


### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake


### Build Configuration

```
force_builders=XCustom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.3.0:20.04
Xbuild_image:Custom Win=openvino-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*

build_image:Custom Win=gapi-onevpl-2021.6.0
buildworker:Custom Win=windows-3
build_contrib:Custom Win=OFF
```

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -30,28 +32,59 @@ namespace gapi {
 
 void VPLDX11AccelerationPolicy::init(session_t session) {
-    mfxStatus sts = MFXVideoCORE_GetHandle(session, MFX_HANDLE_D3D11_DEVICE, reinterpret_cast<mfxHDL*>(&hw_handle));
+    mfxStatus sts = MFXVideoCORE_SetHandle(session, MFX_HANDLE_D3D11_DEVICE, (mfxHDL) hw_handle);
```
`(mfxHDL)`
Why is it not some_cast<>()?

sivanov-work replied:
some kind of copy-paste from public VPL examples. I'll fix it later

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
+        return MFX_ERR_MEMORY_ALLOC;
+    }
+
+    // create  staging texture to read it from
```
```suggestion
    // create staging texture to read it from
```

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,392 @@
+    UINT mapFlags = D3D11_MAP_FLAG_DO_NOT_WAIT;
+
+    shared_device_context->CopySubresourceRegion(get_staging_texture(), 0,
+                                          0, 0, 0,
```
Can this be placed under (.....)?

```
(get_staging_texture(), 0,
 0, 0, 0,
```

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,392 @@
+    GAPI_LOG_DEBUG(nullptr, "record: " << this <<
+                            ", subresources count: " << resources.size());
+
+    for (AllocationId id : resources) {
```
Is this `AllocationId` already reference?

sivanov-work replied:
it is pointer(please find declaration below)

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,392 @@
+    shared_allocator_copy.Alloc = nullptr;
+    shared_allocator_copy.Free = nullptr;
+    shared_allocator_copy.pthis = nullptr;
+
```
Extra space

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,223 @@
+}
+
+VPLMediaFrameDX11Adapter::~VPLMediaFrameDX11Adapter() {
+
```
extra space

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,223 @@
+}
+
+MediaFrame::View VPLMediaFrameDX11Adapter::access(MediaFrame::Access mode) {
+
```
extra space

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,223 @@
+            cv::MediaFrame::View::Ptrs pp = {
+                data.Y,
+                data.UV, nullptr, nullptr
+                };
```
extra tab

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,223 @@
+            cv::MediaFrame::View::Strides ss = {
+                    pitch,
+                    pitch, 0u, 0u
+                };
```
and here 149-151 extra tab

mpashchenkov commented on modules/gapi/src/streaming/onevpl/cfg_params_parser.cpp:
```diff
@@ -67,15 +76,17 @@ std::vector<ValueType> get_params_from_string(const std::string& str) {
-        } else {
+        } else if (name == CfgParam::queue_capacity()) {
+            ret.push_back(creator.create(name, strtoull_or_throw(value.c_str()), false));
+        }else {
```
```suggestion
        } else {
```

mpashchenkov commented on modules/gapi/src/streaming/onevpl/cfg_params_parser.cpp:
```diff
@@ -116,6 +127,32 @@ mfxVariant cfg_param_to_mfx_variant(const CfgParam& cfg_val) {
+    size_t ret = strtoull(str, &end_ptr, 10);
+    if ((end_ptr == str) ||
+        ((ret == LONG_MAX || ret == LONG_MIN) && errno == ERANGE)) {
+            // nothing parsed from the string, handle errors or exit
```
alignment

mpashchenkov commented on modules/gapi/src/streaming/onevpl/cfg_params_parser.cpp:
```diff
@@ -116,6 +127,32 @@ mfxVariant cfg_param_to_mfx_variant(const CfgParam& cfg_val) {
+    int64_t ret = strtoll(str, &end_ptr, 10);
+    if ((end_ptr == str) ||
+        ((ret == LONG_MAX || ret == LONG_MIN) && errno == ERANGE)) {
+            // nothing parsed from the string, handle errors or exit
```
And here?

mpashchenkov commented on modules/gapi/src/streaming/onevpl/source_priv.cpp:
```diff
@@ -211,106 +208,54 @@ GSource::Priv::Priv(std::shared_ptr<IDataProvider> provider,
-                            mfxDecParams.mfx.FrameInfo.Width,
-                            mfxDecParams.mfx.FrameInfo.Height};
-    switch(mfxDecParams.mfx.FrameInfo.FourCC) {
+                            video_param.mfx.FrameInfo.Width,
```
Alignment

mpashchenkov commented on modules/gapi/src/streaming/onevpl/utils.cpp:
```diff
@@ -386,7 +287,7 @@ std::ostream& operator<< (std::ostream& out, const mfxImplDescription& idesc)
     for (int codec = 0; codec < dec.NumCodecs; codec++) {
         auto cid = dec.Codecs[codec].CodecID;
-        out << "(*)mfxImplDescription.mfxDecoderDescription.decoder.CodecID: " << cid;//(cid & 0xff) << "." << (cid >> 8 & 0xff) << "." << (cid >> 16 & 0xff) << "." << (cid >> 24 & 0xff)  << std::endl;
+        out << "(*)" << CfgParam::decoder_id() << ": " << cid;//(cid & 0xff) << "." << (cid >> 8 & 0xff) << "." << (cid >> 16 & 0xff) << "." << (cid >> 24 & 0xff)  << std::endl;
```
Does this debug code make sense?

sivanov-work replied:
no, from prev version. will be removed

mpashchenkov commented on modules/gapi/src/streaming/onevpl/utils.cpp:
```diff
@@ -422,69 +323,65 @@ std::string mfxstatus_to_string(int64_t err) {
+    std::stringstream ss;
+    APPEND_STRINGIFY_MASK_N_ERASE(type, "|", MFX_MEMTYPE_DXVA2_DECODER_TARGET);
+    APPEND_STRINGIFY_MASK_N_ERASE(type, "|", MFX_MEMTYPE_DXVA2_PROCESSOR_TARGET);
+    //APPEND_STRINGIFY_MASK_N_ERASE(type, "|", MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET);
```
Is it debug code? Need to add comment about these if these a future features.

sivanov-work replied:
according to docs MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET has the same digit-like value  as  MFX_MEMTYPE_DXVA2_PROCESSOR_TARGET. Need to put comment about this situation

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    static const size_t inc_count = 10000000;
+    size_t shared_value = 0;
+    auto work = [&lock, &shared_value](size_t count) {
+        for (size_t i = 0; i < count; i ++) {
```
```suggestion
        for (size_t i = 0; i < count; i++) {
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    auto write_work = [&lock, &shared_value](size_t count) {
+        for (size_t i = 0; i < count; i ++) {
+            lock.lock();
+            shared_value ++;
```
```suggestion
            shared_value++;
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    static const size_t inc_count = 10000000;
+    size_t shared_value = 0;
+    auto write_work = [&lock, &shared_value](size_t count) {
+        for (size_t i = 0; i < count; i ++) {
```
```suggestion
        for (size_t i = 0; i < count; i++) {
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    auto read_work = [&lock, &shared_value](size_t count) {
+
+        auto old_shared_value = shared_value;
+        for (size_t i = 0; i < count; i ++) {
```
```suggestion
        for (size_t i = 0; i < count; i++) {
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+
+    // TEST invariant: last_visitor_id MUST be one from any LATTER worker visitor_id
+    bool one_of_available_ids_matched = false;
+    for (size_t id = 0; id < max_worker_count; id ++) {
```
```suggestion
    for (size_t id = 0; id < max_worker_count; id++) {
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+        if (adaptee.owns()) {
+            w_lock_counter ++;
+        } else {
+            r_lock_counter ++;
```
Remove extra space before increment ` ++` 648 and 650

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+        if (adaptee.owns()) {
+            w_unlock_counter ++;
+        } else {
+            r_unlock_counter ++;
```
Remove extra space before increment ` ++` 657 and 659

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+
+    mfxFrameData data;
+    const int exec_count = 123;
+    for (int i = 0; i < exec_count; i ++) {
```
```suggestion
    for (int i = 0; i < exec_count; i++) {
```

mpashchenkov reviewed (commented):
First wave (stupid comments).

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    auto work = [&lock, &shared_value](size_t count) {
+        for (size_t i = 0; i < count; i ++) {
+            lock.lock();
+            shared_value ++;
```
```suggestion
            shared_value++;
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    };
+
+    std::thread worker_thread([&barrier, sync, work] () {
+
```
extra space

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+
+    std::function<mfxStatus(mfxMemId, mfxFrameData *)> lock =
+    [&lock_counter] (mfxMemId, mfxFrameData *) {
+        lock_counter ++;
```
```suggestion
        lock_counter++;
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+
+    mfxFrameData data;
+    const int exec_count = 123;
+    for (int i = 0; i < exec_count; i ++) {
```
```suggestion
    for (int i = 0; i < exec_count; i++) {
```

OrestChura reviewed (commented):
Firs attempt

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.hpp:
```diff
@@ -25,14 +25,15 @@ namespace onevpl {
     void init(session_t session) override;
     void deinit(session_t session) override;
-    pool_key_t create_surface_pool(size_t pool_size, size_t surface_size_bytes, surface_ptr_ctr_t creator) override;
+    pool_key_t create_surface_pool(size_t pool_size, size_t surface_size_bytes, surface_ptr_ctr_t creator);
```
This function now has different parameters than the virtual one has

sivanov-work replied:
it is, it does have different parameters because it doesn't override anything anymore but used as dispatcher

sivanov-work commented on modules/gapi/include/opencv2/gapi/streaming/onevpl/cfg_params.hpp:
```diff
@@ -46,6 +46,62 @@ struct GAPI_EXPORTS CfgParam {
+    /**
+     * @brief queue_capacity
+     *
+     * Special configuration parameter names for onevp::GSource:
```
it is, looks like Doxygen unfold it and make it referenceable. I remember I couldn't navigate otherwise in docs

sivanov-work commented on modules/gapi/include/opencv2/gapi/streaming/onevpl/cfg_params.hpp:
```diff
@@ -46,6 +46,62 @@ struct GAPI_EXPORTS CfgParam {
+     */
+    static constexpr const char *decoder_id() { return "mfxImplDescription.mfxDecoderDescription.decoder.CodecID"; }
+
+    static constexpr const char *implementation() { return "mfxImplDescription.Impl"; }
```
yes. it had better to document it in result.

At now i described only configurable params that have some meaning (Impl can be HARDWARE_ONLY)

sivanov-work commented on modules/gapi/perf/streaming/gapi_streaming_source_perf_tests.cpp:
```diff
@@ -39,12 +42,20 @@ PERF_TEST_P_(OneVPLSourcePerfTest, TestPerformance)
     const auto params = GetParam();
     source_t src = findDataFile(get<0>(params));
     codec_t type = get<1>(params);
+    accel_mode_t mode = get<2>(params);
```
I noticed about tuple. In this specific place I see no benefits still to have rewrite it

OrestChura replied:
After all, it's also about consistency, I believe we're trying to eliminate `.get<>` calls from our tests from time to time

sivanov-work commented on modules/gapi/perf/streaming/gapi_streaming_source_perf_tests.cpp:
```diff
@@ -72,12 +83,17 @@ PERF_TEST_P_(VideoCapSourcePerfTest, TestPerformance)
+                               source_description_t(files[1], codec[1], ""),
+                               source_description_t(files[1], codec[1], "MFX_ACCEL_MODE_VIA_D3D11"),
+                               source_description_t(files[2], codec[2], ""),
+                               source_description_t(files[2], codec[2], "MFX_ACCEL_MODE_VIA_D3D11")));
```
for VPL sources in case of raw video stream input several files depends from codec, so must be pair file <->codec
But Accel Type may be combined with pair file <->codec. I'll try it later

OrestChura replied:
Sorry, I see. Maybe there are no benefits then. Will accept what you decide

sivanov-work commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -271,6 +275,12 @@ int main(int argc, char *argv[]) {
 
+    if (source_queue_capacity != 0) {
+        source_cfgs.push_back(cv::gapi::wip::onevpl::CfgParam::create(
+                                    cv::gapi::wip::onevpl::CfgParam::queue_capacity(),
```
maybe. 
I'll try to apply it

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -21,8 +21,100 @@ namespace cv {
+        case MFX_FOURCC_I010:
+        case MFX_FOURCC_P010:
+            nbytes = width * height + 2 * half_width * half_height;
+            nbytes *= 2;
```
i suppose it was a comment about readability in one of first PRs (this code was moved out from another place)

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
-    throw std::runtime_error("VPLDX11AccelerationPolicy::create_surface_pool() is not implemented");
+    // allocate textures by explicit request
+    mfxFrameAllocResponse mfxResponse;
+    //TODO
```
i think it must be removed

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
+    desc.SampleDesc.Count = 1;
+    desc.Usage = D3D11_USAGE_DEFAULT;
+    desc.MiscFlags = D3D11_RESOURCE_MISC_SHARED;
+    desc.BindFlags = D3D11_BIND_DECODER;
```
it is in 256 line

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
+    mfxHDLPair *pPair = reinterpret_cast<mfxHDLPair *>(handle);
+
+    pPair->first  = data->get_texture();
+    pPair->second = (mfxHDL)reinterpret_cast<DX11AllocationItem::subresource_id_t *>(data->get_subresource());
```
yes, thanks

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,223 @@
+    using stride_t = typename cv::MediaFrame::View::Strides::value_type;
+    stride_t pitch = static_cast<stride_t>(data.Pitch);
+
+    //TODO
```
will be removed

sivanov-work commented on modules/gapi/test/streaming/gapi_streaming_vpl_data_provider.cpp:
```diff
@@ -82,7 +82,7 @@ TEST_P(OneVPL_Source_MFPAsyncDispatcherTest, open_and_decode_file)
     mfx_param_0.Type = MFX_VARIANT_TYPE_U32;
     mfx_param_0.Data.U32 = provider_ptr->get_mfx_codec_id();
-    EXPECT_EQ(MFXSetConfigFilterProperty(cfg_inst_0,(mfxU8 *)"mfxImplDescription.mfxDecoderDescription.decoder.CodecID",
+    EXPECT_EQ(MFXSetConfigFilterProperty(cfg_inst_0,(mfxU8 *)CfgParam::decoder_id(),
```
Sorry, what replacement did you mention?

OrestChura replied:
I mean, when this line became much more shorter, you may consider to squash the next line with it

OrestChura replied:
Not really critical though

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+    LockAdapter adapter(test_allocator.get());
+
+    mfxFrameData data;
+    const int exec_count = 123;
```
Could this magic number become a test parameter in the future?

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_vpl_core_test.cpp:
```diff
@@ -444,6 +509,182 @@ TEST(OneVPL_Source_ProcessingEngine, Init)
+    mfxVariant mfx_param_0;
+    mfx_param_0.Type = MFX_VARIANT_TYPE_U32;
+    mfx_param_0.Data.U32 = MFX_IMPL_TYPE_HARDWARE;
+    EXPECT_EQ(MFXSetConfigFilterProperty(cfg_inst_0,(mfxU8 *)CfgParam::implementation(),
```
`(mfxU8 *)`- c-style cast. I think all c-style casts should be changed to ++ casts.

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    };
+
+    size_t max_worker_count = std::thread::hardware_concurrency();
+    if (max_worker_count < 2) {
```
Does this condition deserve be assert?

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+
+        // TEST invariant: last_visitor_id MUST be one from any FIRST worker visitor_id
+        bool one_of_available_ids_matched = false;
+        for (size_t id = 0; id < total_workers_count; id ++) {
```
```suggestion
        for (size_t id = 0; id < total_workers_count; id++) {
```

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    };
+
+    auto read_work = [&lock, &shared_value](size_t count) {
+
```
empty space

mpashchenkov commented on modules/gapi/test/streaming/gapi_streaming_utils_test.cpp:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    void on_last_out_impl(size_t visitor_id) {
+
```
Empty space

mpashchenkov commented on modules/gapi/test/common/gapi_streaming_tests_common.hpp:
```diff
@@ -0,0 +1,91 @@
+};
+
+static const unsigned char hevc_header[] = {
+ 0x00, 0x00, 0x00, 0x01, 0x40, 0x01, 0x0C, 0x06, 0xFF, 0xFF, 0x01, 0x40, 0x00,
+ 0x00, 0x03, 0x00, 0x80, 0x00, 0x00, 0x03, 0x00, 0x00, 0x03, 0x00, 0x78, 0x00,
+ 0x00, 0x04, 0x02, 0x10, 0x30, 0x00, 0x00, 0x03, 0x00, 0x10, 0x00, 0x00, 0x03,
+ 0x01, 0xE5, 0x00, 0x00, 0x00, 0x01, 0x42, 0x01, 0x06, 0x01, 0x40, 0x00, 0x00,
+ 0x03, 0x00, 0x80, 0x00, 0x00, 0x03, 0x00, 0x00, 0x03, 0x00, 0x78, 0x00, 0x00,
+ 0xA0, 0x10, 0x20, 0x61, 0x63, 0x41, 0x00, 0x86, 0x49, 0x1B, 0x2B, 0x20, 0x00,
+ 0x00, 0x00, 0x01, 0x44, 0x01, 0xC0, 0x71, 0xC0, 0xD9, 0x20, 0x00, 0x00, 0x00,
+ 0x01, 0x26, 0x01, 0xAF, 0x0C
```
One tab instead one space.

mpashchenkov commented on modules/gapi/src/streaming/onevpl/utils.cpp:
```diff
@@ -14,160 +14,114 @@
+    ONEVPL_STRINGIFY_IF(MFX_ACCEL_MODE_VIA_VAAPI_WAYLAND)
+    ONEVPL_STRINGIFY_IF(MFX_ACCEL_MODE_VIA_HDDLUNITE)
+    throw std::logic_error(std::string("Invalid \"") +  CfgParam::acceleration_mode() +
+                                       "\":" + cstr);
```
Please move this under (**S**td::string("I...

mpashchenkov commented on modules/gapi/src/streaming/onevpl/cfg_params_parser.cpp:
```diff
@@ -44,6 +41,18 @@ struct ParamCreator<mfxVariant> {
+        ret.Data.I64 = value;
+        return ret;
+    }
+    mfxVariant create_impl(const std::string&, mfxU64 value) {
```
Why are create_impl functions not a template? Is only three mfx-types expected? Does template bring only difficulties?

mpashchenkov commented on modules/gapi/src/streaming/onevpl/accelerators/utils/shared_lock.cpp:
```diff
@@ -0,0 +1,96 @@
+    bool pred_excl = exclusive_lock.load();
+    do {
+        if (!pred_excl) {
+            // if no exclusive lock than start shared lock transaction
```
th**a**n --> th**e**n

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
+    if (pool_it == pool_table.end()) {
+        std::stringstream ss;
+        ss << "key is not found: " << key << ", table size: " << pool_table.size();
+        const std::string& str = ss.str();
```
You are using "Type& var" and "Type &var" from time to time across this file (and maybe others too)
It is a really minor suggestion, but you may consider to align it :)

OrestChura replied:
I hope it'll be fixed in the future

OrestChura reviewed (commented):
№ 2

anna-khakimova commented:
Looks good for me. Great work!

alalek commented:
Please configure builders with enabled oneVPL support (see previous oneVPL-related PRs)

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,416 @@
+#include <CL/cl_d3d11.h>
+#endif // HAVE_OPENCL
+#undef D3D11_NO_HELPERS
+#undef NOMINMAX
```
Is it necessary to have this preproc. logic even when it is already included in "streaming/onevpl/accelerators/dx11_alloc_resource.hpp" ?

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.hpp:
```diff
@@ -0,0 +1,151 @@
+struct GAPI_EXPORTS LockAdapter {
+    LockAdapter(mfxFrameAllocator origin_allocator);
+
+    // GAPI_EXPORTS for tests
```
Is it the right place for this comment?
I'd prefer it above `struct GAPI_EXPORTS LockAdapter`

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.hpp:
```diff
@@ -0,0 +1,151 @@
+
+    void write_lock(mfxMemId mid, mfxFrameData &data);
+    bool is_write_acquired();
+    void unlock_write(mfxMemId mid, mfxFrameData &data);
```
Just for naming - why `smth_lock` but `unlock_smth`?

OrestChura replied:
OK, let it be, I agree :)

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,416 @@
+    return MFX_ERR_NONE;
+}
+
+DX11AllocationRecord::DX11AllocationRecord() = default;
```
can this be declared in `.hpp`?

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.hpp:
```diff
@@ -0,0 +1,151 @@
+        std::shared_ptr<DX11AllocationRecord> record(new DX11AllocationRecord);
+        record->init(std::forward<Args>(args)...);
+        return record;
+    }
```
Should this implementation be moved in `.cpp`?

OrestChura replied:
OK.. is it just to edit 1 file instead of 2 in such case?  
I don't think it is a good practice, like, what if it will **not** be expanded? then this implementation sticks here, forgotten

OrestChura replied:
I see now. OK, my bad

OrestChura commented on modules/gapi/include/opencv2/gapi/streaming/onevpl/cfg_params.hpp:
```diff
@@ -46,9 +46,72 @@ struct GAPI_EXPORTS CfgParam {
+     * Create generic onevp::GSource configuration parameter.
      *
      *@param name           name of parameter.
      *@param value          value of parameter.
```
Should this `create()` function implementation be in `cfg_params.cpp` file?

sivanov-work commented on modules/gapi/src/streaming/onevpl/utils.hpp:
```diff
@@ -42,11 +42,19 @@ void release(COMNonManageableType *ptr) {
+template <typename COMNonManageableType>
+ComSharedPtrGuard<COMNonManageableType> createCOMSharedPtrGuard(ComPtrGuard<COMNonManageableType>&& unique_guard) {
+    return ComSharedPtrGuard<COMNonManageableType>(std::move(unique_guard));
+}
```
it's templated

sivanov-work commented on modules/gapi/test/common/gapi_streaming_tests_common.hpp:
```diff
@@ -0,0 +1,91 @@
+        EXPECT_TRUE(in);
+    }
+
+mfx_codec_id_type get_mfx_codec_id() const override {
```
unit tests, i think we can skip that style checks in tests :)

OrestChura commented on modules/gapi/CMakeLists.txt:
```diff
@@ -275,12 +275,19 @@ if(HAVE_GAPI_ONEVPL)
+  target_compile_options(${the_module} PUBLIC "/wd4201"
+                                              "/wd4302"
+                                              "/wd4311"
+                                              "/wd4312")
```
I don't think it is good to suppress these warnings for all the module. Is it because of oneVPL headers includes?
Then we should maybe cover it with `pragma push-pop`s, not to affect all the other compiler code check

OrestChura replied:
Great!

OrestChura commented on modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:
```diff
@@ -242,9 +242,16 @@ VPLLegacyDecodeEngine::initialize_session(mfxSession mfx_session,
-        decRequest.NumFrameSuggested = preallocated_frames_count;
+        if (static_cast<size_t>(std::numeric_limits<mfxU16>::max()) < preallocated_frames_count) {
+            GAPI_LOG_WARNING(nullptr, "Cannot proceed with CfgParam \"" << CfgParam::frames_pool_size_name() << "\": " <<
+                                  preallocated_frames_count << ". It must not be equal than " <<
```
`must not be greater` I believe

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
+    }
+
+    pool_t& requested_pool = pool_it->second;
+    return requested_pool.find_free();
```
OK, if you think it would be easier that way

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -63,53 +96,288 @@ void VPLDX11AccelerationPolicy::deinit(session_t session) {
+    VPLDX11AccelerationPolicy *self = static_cast<VPLDX11AccelerationPolicy *>(pthis);
+    GAPI_LOG_DEBUG(nullptr, "called from: " << self ? "Policy" : "Resource");
+    cv::util::suppress_unused_warning(self);
+    return on_unlock(mid, ptr);
```
OK. I didn't actually get it deeply, but I understood about life scopes and that there are differences between `alloc_cb` and `lock_cb`, for example. I believe you've got thoughts on that one more time while describing, so I must be OK :)

OrestChura commented on modules/gapi/CMakeLists.txt:
```diff
@@ -272,12 +275,16 @@ if(HAVE_GAPI_ONEVPL)
     ocv_target_link_libraries(opencv_test_gapi PRIVATE ${VPL_IMPORTED_TARGETS})
+    if(MSVC)
+        target_compile_options(opencv_test_gapi PUBLIC "/wd4201")
+    endif()
```
You've rewritten includes via `pragma` in tests, too - is it still necessary?

OrestChura commented on modules/gapi/test/common/gapi_streaming_tests_common.hpp:
```diff
@@ -0,0 +1,91 @@
+        if (!out_bitstream) {
+            out_bitstream = std::make_shared<mfx_bitstream>();
+            out_bitstream->MaxLength = 2000000;
+            out_bitstream->Data = (mfxU8 *)calloc(out_bitstream->MaxLength, sizeof(mfxU8));
```
Resolving as not relevant, unit tests

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.hpp:
```diff
@@ -39,30 +36,52 @@ namespace onevpl {
+    using allocation_t = std::shared_ptr<DX11AllocationRecord>;
+    std::map<alloc_id_t, allocation_t> allocation_table;
+
+    std::map<pool_key_t, pool_t> pool_table;
```
Related to https://github.com/opencv/opencv/pull/21049#discussion_r761817874

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/utils/shared_lock.cpp:
```diff
@@ -0,0 +1,96 @@
+            std::this_thread::yield();
+        }
+        prev_shared = shared_counter.load();
+    } while (prev_shared != 0 || !in_progress);
```
That's what I meant - in the end, `in_progress == true`, but you must proceed further. The comment is regarding naming. "In progress" itself means you are in the middle of some process and you cannot move further. It doesn't match with the meaning of this variable

If to suggest, this variable means that write access is gained, so it can be named like `gained`, `succeeded`; or, better, meaningful `locked`, because `exclusive_lock` **was** set true, but in the case you've described it should be rolled back. But it is `locked` that time.

sivanov-work replied:
this variable cannot be named as `gained`, `succeeded` of `locked` because resource is not locked until we finished final `while`. It actually means: `we are in the middle of progress something` and we must not make  some assumption about succeed or not until we check the second condition  `shared_lock ==0`

So `in_progress ` means: I increased exclusive lock counter but i'm not sure that i'm allowed to continue with exclusive access until i check the second condition.
This is sync with using two independent atomic states: and we cannot proceed it both of them wouldn't reach the right state

OrestChura commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,416 @@
+
+size_t LockAdapter::unlock_read(mfxMemId mid, mfxFrameData &data) {
+    GAPI_DbgAssert(!impl || !is_write_acquired() &&
+                   "Reject `unlock_read` in `write_lock` state");
```
Do I understand you correctly? If there is a write-lock, there can't be a read thread so this `unlock_read` must be rejected (at least while debugging).

but there is negation in the condition. `!is_write_acqired()` means "Write-lock is **not** acquired". Is it the condition you've meant for this Assert?

OrestChura replied:
OK, I got it, thanks! I confused what should be under the Assert, sorry

Still, there is something - what is the first part `!impl` for? Having or not having `impl` is checked in `is_write_acquired` itself, why checking it here in the first place?
And then, it seems not right for me. `impl` must be present for `read_lock` being possible as you've explained, so there should be `Assert(impl)`. But here under Assert is `!impl`, the opposite.

sivanov-work replied:
there is some trick part:

when no `impl` - then it means that all read/write exposed as `write` (with the  most strongest guarantee).
So no matter if you invoke read or write or then unlock_write, unlock_read. So `is_write_acquired` must return `true` if no `impl`

But when adaptee `impl` is set, then it require different logic and `is_write_acquired` must return actual type when impl is exist

OrestChura reviewed (approved):
Thank you for all the efforts!

mpashchenkov reviewed (approved):
I didn't find places with possible nulls and going out of memory. I'm give up.

alalek commented on modules/gapi/src/streaming/onevpl/onevpl_export.hpp:
```diff
@@ -0,0 +1,25 @@
+#ifndef GAPI_STREAMING_ONEVPL_EXPORT_HPP
+#define GAPI_STREAMING_ONEVPL_EXPORT_HPP
+
+#if defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable : 4201)
+#pragma warning(disable : 4302)
+#pragma warning(disable : 4311)
+#pragma warning(disable : 4312)
+#endif // defined(_MSC_VER)
```
Because we should not disable warnings anywhere.
Problems with warnings should be suppressed locally.

sivanov-work commented:
if no other comments are coming then I assume to apply them in the next PRs:
- warning for cmake disabled for onevpl in unit tests only, regular build is unaffected: move them into header inclusion in tests
- some typos in log & assert infos
- to rename shared_lock : `in_progress` to `is_pending` etc.

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,404 @@
+
+    // dispatch to VPL allocator using READ access mode
+    mfxStatus sts = MFX_ERR_LOCK_MEMORY;
+    try {
+        sts = lockable_allocator.Lock(nullptr, mid, &data);
+    } catch(...) {
+    }
```
???

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,404 @@
+    } catch(...) {
+    }
+
+    // adapter will throw error if VPL frame allocator fails
```
If it throws then you just caught and discarded it in the line above. Why just not handle it with RAII lock over `impl` here?

sivanov-work replied:
yes, some ugly code below, but cover both cases for exception and regular fail

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,404 @@
+    if (sts != MFX_ERR_NONE) {
+        impl->unlock_shared();
+        GAPI_Assert(false && "Cannot lock frame on READ using VPL allocator");
+    }
```
Logically this code is also ok but may be harder to understand

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -0,0 +1,404 @@
+
+SharedLock* LockAdapter::get_adaptee() {
+    return impl;
+}
```
So if you have an `rwlock` here, can this structure be generalized or why it is tightly bound to `mfx`?

sivanov-work replied:
it adaptee MediaFrameAdapter locking by R/W  to mfx callback on_lock/on_unlock in DX11AccelPolicy 

This adapter change locking behavior. when it is not set, then callback from VPL decode on_lock/on_unlock behaves  as exclusive mutex, but when DX11MediaAdapter is created then on_lock/ob_unlock would work as shared mutex

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,232 @@
+    LockAdapter* alloc_data = reinterpret_cast<LockAdapter *>(mid);
+    if (mode == MediaFrame::Access::R) {
+        alloc_data->read_lock(mid, data);
+    } else {
```
since the new types of `Access` may be introduced in the future (e.g., I do consider `RW` to be the case), this `else` construct should be more robust I believe (or better to switch to `switch` already)

sivanov-work replied:
ok, agree.
TODO in next PR

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,232 @@
+    LockAdapter* alloc_data = reinterpret_cast<LockAdapter*>(data.MemId);
+    if (mode == MediaFrame::Access::R) {
+        alloc_data->unlock_read(mid, data);
+    } else {
```
Same issue here

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp:
```diff
@@ -0,0 +1,232 @@
+    if (mode == MediaFrame::Access::R) {
+        alloc_data->unlock_read(mid, data);
+    } else {
+        alloc_data->unlock_write(mid, data);
```
so can this be a plain abstract `rwlock` (e.g. like one from boost or pthreads)? Or something differs in the access at the VPL level?

sivanov-work replied:
adapter is quite different, but shared_mutex inside it should be reused from official sources further
P.S. `is_own_locked`  can be put on upper level to do not customize  official shared_mutex
P.P.S. official c++17 shared_mutex is mutex-based with wait-lock, but tbb has imple,entation with rw_**spinlock** to make busy wait

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/utils/elastic_barrier.hpp:
```diff
@@ -0,0 +1,296 @@
+namespace onevpl {
+
+template<typename Impl>
+class elastic_barrier {
```
Is the idea of this structure described anywhere?

dmatveev replied:
(I'd expect to see it here)

sivanov-work replied:
May i add in the next PR such description?
In short overview: 
pass only single one thread (the first one) to make some initialization: resource acquisition - another will waiting for the first thread job completion.
And make final deinitialization: resource release - only for last one( if no other threads hold resource or intendent to acquire it)

Some kind of double check lock patter in Gstreamer MediaResource Adapter - but support more scenarios( and more efficient?)

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/utils/shared_lock.cpp:
```diff
@@ -0,0 +1,95 @@
+}
+bool SharedLock::owns() const {
+    return exclusive_lock.load();
+}
```
why do we need our own shared lock here?

sivanov-work replied:
yes it is, but this can be easily emulated on upper level in LockAdapter

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/utils/shared_lock.hpp:
```diff
@@ -0,0 +1,47 @@
+namespace wip {
+namespace onevpl {
+
+class GAPI_EXPORTS SharedLock {
```
is it modelled after the C++14's shared_lock?

sivanov-work replied:
yes, homegrown solution...

dmatveev commented on modules/gapi/src/streaming/onevpl/cfg_params_parser.cpp:
```diff
@@ -116,6 +127,32 @@ mfxVariant cfg_param_to_mfx_variant(const CfgParam& cfg_val) {
     return ret;
 }
+
+size_t strtoull_or_throw(const char* str) {
```
y use chars

sivanov-work replied:
to satisty c-string in  interface functions (to make interface ABI consistent)

dmatveev reviewed (approved):
Approved with comments

sivanov-work commented on modules/gapi/perf/streaming/gapi_streaming_source_perf_tests.cpp:
```diff
@@ -39,12 +42,20 @@ PERF_TEST_P_(OneVPLSourcePerfTest, TestPerformance)
     };
 
+    if (!type.empty()) {
+        cfg_params.push_back(CfgParam::create_decoder_id(type.c_str()));
```
to retain public interface consistent (maybe for future): because STL classes are not ABI compatible in general

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -21,8 +21,100 @@ namespace cv {
 namespace wip {
 namespace onevpl {
+namespace utils {
+mfxU32 GetSurfaceSize_(mfxU32 FourCC, mfxU32 width, mfxU32 height) {
```
it was get from VPL examples, may be we could add it when faces  with the issue later

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -21,8 +21,100 @@ namespace cv {
+    return nbytes;
+}
+
+surface_ptr_t create_surface_RGB4_(mfxFrameInfo frameInfo,
+                                   std::shared_ptr<void> out_buf_ptr,
+                                   size_t out_buf_ptr_offset,
+                                   size_t out_buf_size)
```
it is shared across surfaces: each surface memory is sliced from `out_buf_ptr`
So all surfaces have reference to that buf and mapped on that buffer memory  by offset

Referencing is required to keep `out_buf_ptr`  alive when alive at least one surface

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -21,8 +21,100 @@ namespace cv {
+    // TODO more intelligent check
+    if (out_buf_size <=
+        out_buf_ptr_offset + (surfW * surfH) + ((surfW / 2) * (surfH / 2))) {
+        GAPI_LOG_WARNING(nullptr, "Not enough buffer, ptr: " << out_buf_ptr <<
```
TODO in next PR

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.hpp:
```diff
@@ -25,14 +24,15 @@ namespace onevpl {
 struct GAPI_EXPORTS VPLCPUAccelerationPolicy final : public VPLAccelerationPolicy
 {
-    VPLCPUAccelerationPolicy();
+    VPLCPUAccelerationPolicy(device_selector_ptr_t selector);
```
TODO in next PR

sivanov-work commented:
@alalek  Could you merge it please? Non addressed comment would be addressed at the next PR
Custom build for Win+gapi-onevpl  is passed (warning fixed and disarmed for UT)
